src/common/base/src/runtime/runtime.rs

```diff
@@ -325,3 +326,32 @@ pub async fn match_join_handle<T>(handle: JoinHandle<Result<T>>) -> Result<T> {
         },
     }
 }
+
+/// Run multiple futures parallel
+/// using a semaphore to limit the parallelism number, and a specified thread pool to run the futures.
+/// It waits for all futures to complete and returns their results.
+pub async fn execute_futures_in_parallel<Fut>(
+    futures: impl IntoIterator<Item = Fut>,
+    thread_nums: usize,
+    permit_nums: usize,
+    thread_name: String,
+) -> Result<Vec<Fut::Output>>
+where
+    Fut: Future + Send + 'static,
+    Fut::Output: Send + 'static,
+{
+    // 1. build the runtime.
+    let semaphore = Semaphore::new(permit_nums);
+    let runtime = Arc::new(Runtime::with_worker_threads(
```

Comment:
Seems it will panic with error: `Cannot start a runtime from within a runtime`

Reply:
We have a DIY runtime(by @drmingdrmer has a better understanding of it and can help explain it), this is not the same as the origin tokio:
https://github.com/datafuselabs/databend/blob/d28aa6ab9d7a5f2f22e0ab4cb375479d379b5d7a/src/common/base/src/runtime/runtime.rs

And this PR is not the first time to use as this :)